subprojects/core-api/src/main/java/org/gradle/api/provider/Provider.java

```diff
@@ -135,6 +172,11 @@ public interface Provider<T> {
      * will also have no value and the spec will not be called.
      * </p>
      *
+     * <h4>Configuration cache</h4>
+     * <p>
+     * The resulting provider is evaluated at store time if this provider can be evaluated at store time.
```

Comment:
> if this provider can be evaluated at store time

🤔  I find this part potentially introducing more questions. We should provide a link or a sentence on how to (reasonably) find out whether the current provider can be evaluated. I understand it can only be truly determined at runtime, but the javadoc reader should get a pointer to some overview maybe.

> can be evaluated at store time

I want to be very nit-picky about the language here, but it may not be as important.

Couldn't all providers be evaluated at configuration time? The way we draw the line is more the opposite -- "can the provider be NOT evaluated at configuration time", and if we can avoid it, we don't evaluate it. And we can avoid it iff we know that the provider is not used at configuration time and its evaluation is not going to introduce configuration inputs we want to observe.

The logic may be clearer for the reader if we explain it via the reason why the provider might be evaluated _earlier_ at the store time with CC as opposed to being evaluated _later_ at task input snapshotting time in vintage.

It might not be reasonable to explain this in the javadoc of every method, but instead have a direct link to that shared explanation.